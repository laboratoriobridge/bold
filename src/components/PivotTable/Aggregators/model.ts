import { LocaleConfiguration } from '../../../i18n/LocaleContext'

export enum AggregatorEnum {
  PERCENTAGE = 'PERCENTAGE',
  COUNT = 'COUNT',
  AVERAGE = 'AVERAGE',
  MAXIMUM = 'MAXIMUM',
  MINIMUM = 'MINIMUM',
}

export type AggregatorLabels = Pick<
  LocaleConfiguration['aggregators'],
  'count' | 'percentage' | 'average' | 'maximum' | 'minimum'
>

export type AggregatorFunction = (values: number[], total?: number) => number

export type Aggregator = {
  id: AggregatorEnum
  label: string
  value: AggregatorFunction | undefined
  keyDependent: boolean
  chain?: Aggregator[]
  suffix?: string
}
