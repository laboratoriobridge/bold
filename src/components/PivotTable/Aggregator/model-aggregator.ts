export enum AggregatorEnum {
  PERCENTAGE = 'PERCENTAGE',
  COUNT = 'COUNT',
  AVERAGE = 'AVERAGE',
  MAXIMUM = 'MAXIMUM',
  MINIMUM = 'MINIMUM',
}

export type AggregatorFunction = (values: number[], total?: number) => number

export type Aggregator = {
  id: AggregatorEnum
  label: string
  value: AggregatorFunction | undefined
  keyDependent: boolean
  chain?: Aggregator[]
  suffix?: string
}
