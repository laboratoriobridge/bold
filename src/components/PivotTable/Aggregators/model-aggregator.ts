export enum AggregatorEnum {
  PERCENTAGE = 'PERCENTAGE',
  COUNT = 'COUNT',
  AVERAGE = 'AVERAGE',
  MAXIMUM = 'MAXIMUM',
  MINIMUM = 'MINIMUM',
}

export interface AggregatorLabels {
  count: string
  percentage: string
  average: string
  maximum: string
  minimum: string
}

export type AggregatorFunction = (values: number[], total?: number) => number

export type Aggregator = {
  id: AggregatorEnum
  label: string
  value: AggregatorFunction | undefined
  keyDependent: boolean
  chain?: Aggregator[]
  suffix?: string
}
