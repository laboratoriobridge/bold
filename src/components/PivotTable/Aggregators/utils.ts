import { Aggregator, AggregatorEnum, AggregatorLabels } from './model'

export const getCountAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.COUNT,
  label: aggregatorLabels.count,
  value: undefined,
  keyDependent: false,
})

const getPercentageAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.PERCENTAGE,
  label: aggregatorLabels.percentage,
  value: (values: number[], total: number): number => (values.reduce((prev, curr) => prev + curr, 0) * 100) / total,
  keyDependent: false,
  chain: [getCountAggregator(aggregatorLabels)],
  suffix: '%',
})

const getAverageAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.AVERAGE,
  label: aggregatorLabels.average,
  value: (values: number[]): number => values.reduce((prev, curr) => prev + curr, 0) / values.length,
  keyDependent: true,
})

const getMaximumAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.MAXIMUM,
  label: aggregatorLabels.maximum,
  value: (values: number[]): number => values.reduce((prev, curr) => (prev > curr ? prev : curr)),
  keyDependent: true,
})

const getMinimumAggregator = (aggregatorLabels: AggregatorLabels): Aggregator => ({
  id: AggregatorEnum.MINIMUM,
  label: aggregatorLabels.minimum,
  value: (values: number[]): number => values.reduce((prev, curr) => (prev < curr ? prev : curr)),
  keyDependent: true,
})

export const getAggregators = (aggregatorLabels: AggregatorLabels) => {
  return [
    getCountAggregator(aggregatorLabels),
    getPercentageAggregator(aggregatorLabels),
    getAverageAggregator(aggregatorLabels),
    getMaximumAggregator(aggregatorLabels),
    getMinimumAggregator(aggregatorLabels),
  ]
}

export const getKeyNotDependentAggregators = (aggregators: Aggregator[]) => {
  return aggregators.filter((agg) => !agg.keyDependent)
}
