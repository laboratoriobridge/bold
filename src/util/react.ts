import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
