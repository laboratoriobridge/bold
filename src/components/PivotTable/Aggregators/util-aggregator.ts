import { partition } from 'lodash'

import { Aggregator, AggregatorEnum } from './model-aggregator'

export const getCountAndPercentageAggregator = (countLabel, percentageLabel) => {
  const COUNT_AGGREGATOR: Aggregator = {
    id: AggregatorEnum.COUNT,
    label: countLabel,
    value: undefined,
    keyDependent: false,
  }

  const PERCENTAGE_AGGREGATOR: Aggregator = {
    id: AggregatorEnum.PERCENTAGE,
    label: percentageLabel,
    value: (values: number[], total: number): number => (values.reduce((prev, curr) => prev + curr, 0) * 100) / total,
    keyDependent: false,
    chain: [COUNT_AGGREGATOR],
    suffix: '%',
  }

  return { COUNT_AGGREGATOR, PERCENTAGE_AGGREGATOR }
}

export const getAggregators = (countLabel, percentageLabel, averageLabel, maximumLabel, minimumLabel) => {
  const { COUNT_AGGREGATOR, PERCENTAGE_AGGREGATOR } = getCountAndPercentageAggregator(countLabel, percentageLabel)

  return [
    COUNT_AGGREGATOR,
    PERCENTAGE_AGGREGATOR,
    {
      id: AggregatorEnum.AVERAGE,
      label: averageLabel,
      value: (values: number[]): number => values.reduce((prev, curr) => prev + curr, 0) / values.length,
      keyDependent: true,
    },
    {
      id: AggregatorEnum.MAXIMUM,
      label: maximumLabel,
      value: (values: number[]): number => values.reduce((prev, curr) => (prev > curr ? prev : curr)),
      keyDependent: true,
    },
    {
      id: AggregatorEnum.MINIMUM,
      label: minimumLabel,
      value: (values: number[]): number => values.reduce((prev, curr) => (prev < curr ? prev : curr)),
      keyDependent: true,
    },
  ]
}

export const getKeyDependentAndNotDependentAggregators = (AGGREGATORS: Aggregator[]) => {
  return partition(AGGREGATORS, (agg) => agg.keyDependent)
}
