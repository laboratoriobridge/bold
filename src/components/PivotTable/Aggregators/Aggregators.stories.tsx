import React, { useState } from 'react'

import { optionsKnob } from '@storybook/addon-knobs'
import { useLocale } from '../../../i18n'
import { Aggregators } from './Aggregators'
import { Aggregator, KeyMap } from './model-aggregator'
import { getCountAndPercentageAggregator } from './util-aggregator'

export default {
  title: 'Components/PivotTable/Aggregators',
}

type Fruit = {
  name: string
  size: string
}

const keyMapping: KeyMap<Fruit> = new Map([
  ['name', { keyName: 'Name' }],
  ['size', { keyName: 'Size' }],
])

const numberKeyOptions = {
  Empty: [],
  Name: ['name'],
  'Name and Size': ['name', 'size'],
}

export const Default = () => {
  const locale = useLocale()

  const initialAggregator = getCountAndPercentageAggregator(locale.aggregators.COUNT, locale.aggregators.PERCENTAGE)
    .COUNT_AGGREGATOR

  const [aggregator, setAggregator] = useState<Aggregator>(initialAggregator)

  const [aggregatorKey, setAggregatorKey] = useState<keyof Fruit>()

  const handleAggregatorChange = (aggregator: Aggregator) => {
    setAggregator(() => aggregator)
  }

  const handleAggregatorKeyChange = (key: keyof Fruit) => setAggregatorKey(key)

  const numberKeys = optionsKnob('Number Keys', numberKeyOptions, [], { display: 'select' }) as Array<keyof Fruit>

  return (
    <Aggregators<Fruit>
      numberKeys={numberKeys}
      keyMap={keyMapping}
      handleAggregatorChange={handleAggregatorChange}
      handleAggregatorKeyChange={handleAggregatorKeyChange}
      aggregator={aggregator}
      aggregatorKey={aggregatorKey}
    />
  )
}
