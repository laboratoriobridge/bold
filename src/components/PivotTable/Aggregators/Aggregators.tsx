/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import React from 'react'

import { Radio } from '../../Radio'
import { Select } from '../../Select'
import { VFlow } from '../../VFlow'
import { Aggregator } from './model-aggregator'
import { AGGREGATORS, KEY_NOT_DEPENDENT_AGGREGATORS, KEY_DEPENDENT_AGGREGATORS } from './util-aggregator'

export type AggregatorsProps<T extends any> = {
  numberKeys: string[]
  keyMapping: Map<
    keyof T,
    { keyName: string; formatter?: (value: string) => string; ordenator?: (a: string, b: string) => number }
  >
  handleAggregatorChange: (aggregator: Aggregator) => void
  handleAggregatorKeyChange: (key: keyof T) => void
  aggregator: Aggregator
  aggregatorKey: keyof T
}

export function Aggregators<T extends any>(props: AggregatorsProps<T>) {
  const { numberKeys, keyMapping, handleAggregatorKeyChange, handleAggregatorChange, aggregator, aggregatorKey } = props

  const styles = createStyles()

  const itemToString = (item: keyof T | null) => (item ? keyMapping.get(item).keyName || (item as string) : '')

  const handleKeySelect = (item: keyof T) => {
    handleAggregatorKeyChange(item)
  }

  const handleAggregatorSelect = (evt: React.ChangeEvent<HTMLInputElement>) => {
    const idx = Number(evt.target.value)
    handleAggregatorChange(AGGREGATORS[idx])
  }

  const numberKeysIsEmpty = numberKeys.length === 0

  return (
    <VFlow>
      <div css={styles.container}>
        {(numberKeysIsEmpty ? KEY_NOT_DEPENDENT_AGGREGATORS : AGGREGATORS).map((f, idx) => (
          <div key={f.id} css={styles.wrapper}>
            <Radio
              name='aggregator'
              checked={aggregator.id === f.id}
              disabled={f.keyDependent && numberKeysIsEmpty}
              label={f.label}
              value={idx}
              onChange={handleAggregatorSelect}
            />
          </div>
        ))}
      </div>
      {KEY_DEPENDENT_AGGREGATORS.includes(aggregator) && (
        <Select<keyof T>
          disabled={numberKeysIsEmpty}
          items={numberKeys as Array<keyof T>}
          itemToString={itemToString}
          value={aggregatorKey}
          onChange={handleKeySelect}
        />
      )}
    </VFlow>
  )
}

const createStyles = () => ({
  container: css`
    display: flex;
    flex-wrap: wrap;
    justify-content: flex-start;
  `,
  wrapper: css`
    padding-top: 1rem;
    padding-left: 1rem;
    margin-bottom: 0.5rem;
    margin-right: 1rem;
  `,
})
