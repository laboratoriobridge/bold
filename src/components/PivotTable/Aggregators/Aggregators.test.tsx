import { render } from '@testing-library/react'
import React from 'react'
import { Aggregators } from './Aggregators'
import { Aggregator, AggregatorEnum, KeyMap } from './model-aggregator'

describe('PivotTable - Aggregators', () => {
  type Test = {
    name: string
  }

  const handleAggregator = (aggregator: Aggregator) => null

  const handleAggregatorKey = (key: keyof Test) => null

  const keyMapping: KeyMap<Test> = new Map([['name', { keyName: 'Name' }]])

  const TEST_AGGREGATOR: Aggregator = {
    id: AggregatorEnum.COUNT,
    label: 'Test',
    value: undefined,
    keyDependent: false,
  }

  it('should render correctly', () => {
    const { container } = render(
      <Aggregators<Test>
        numberKeys={[]}
        keyMap={keyMapping}
        handleAggregatorChange={handleAggregator}
        handleAggregatorKeyChange={handleAggregatorKey}
        aggregator={TEST_AGGREGATOR}
        aggregatorKey={'name'}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('should render correctly with numberKeys', () => {
    const { container } = render(
      <Aggregators<Test>
        numberKeys={['name']}
        keyMap={keyMapping}
        handleAggregatorChange={handleAggregator}
        handleAggregatorKeyChange={handleAggregatorKey}
        aggregator={TEST_AGGREGATOR}
        aggregatorKey={'name'}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
