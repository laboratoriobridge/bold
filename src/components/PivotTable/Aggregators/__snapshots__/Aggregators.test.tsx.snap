// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PivotTable - Aggregators should render correctly 1`] = `
.emotion-12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-11 {
  display: inline-block;
  margin-bottom: 0.5rem;
  margin-top: 0.5rem;
}

.emotion-11:first-of-type {
  margin-top: 0;
}

.emotion-11:last-of-type {
  margin-bottom: 0;
}

.emotion-11:empty {
  display: none;
}

.emotion-3 {
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  background-color: #FFFFFF;
  border: 1px solid #8F8FA2;
  border-radius: 100px;
  display: inline-block;
  height: 24px;
  position: relative;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  vertical-align: middle;
  width: 24px;
}

.emotion-1:after {
  background-color: #FFFFFF;
  border: 5px solid #FFFFFF;
  border-radius: 100px;
  content: "";
  display: block;
  height: 2px;
  margin-left: 6px;
  margin-top: 6px;
  opacity: 0;
  text-align: center;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  width: 2px;
}

.emotion-2 {
  margin-left: 0.5rem;
}

.emotion-0 {
  opacity: 0;
  margin-right: -13px;
}

.emotion-0:hover + .emotion-1 {
  border-color: #7A7B8E;
}

.emotion-0:checked + .emotion-1 {
  background-color: #0069D0;
  border-color: #0069D0;
}

.emotion-0:checked + .emotion-1:after {
  opacity: 1;
}

.emotion-0:focus + .emotion-1 {
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-0:disabled + .emotion-1 {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-0:disabled + .emotion-1 + .emotion-2 {
  color: #AAAAB9;
  cursor: not-allowed;
}

.emotion-10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.emotion-4 {
  padding-top: 1rem;
  padding-left: 1rem;
  margin-bottom: 0.5rem;
  margin-right: 1rem;
}

<div>
  <div
    class="emotion-12"
  >
    <div
      class="emotion-11"
    >
      <div
        class="emotion-10"
      >
        <div
          class="emotion-4"
        >
          <label
            class="emotion-3"
          >
            <input
              checked=""
              class="emotion-0"
              name="aggregator"
              type="radio"
              value="0"
            />
            <span
              class="emotion-1"
            />
            <span
              class="emotion-2"
            >
              Count
            </span>
          </label>
        </div>
        <div
          class="emotion-4"
        >
          <label
            class="emotion-3"
          >
            <input
              class="emotion-0"
              name="aggregator"
              type="radio"
              value="1"
            />
            <span
              class="emotion-1"
            />
            <span
              class="emotion-2"
            >
              Percentage
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PivotTable - Aggregators should render correctly with numberKeys 1`] = `
.emotion-27 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-26 {
  display: inline-block;
  margin-bottom: 0.5rem;
  margin-top: 0.5rem;
}

.emotion-26:first-of-type {
  margin-top: 0;
}

.emotion-26:last-of-type {
  margin-bottom: 0;
}

.emotion-26:empty {
  display: none;
}

.emotion-3 {
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  background-color: #FFFFFF;
  border: 1px solid #8F8FA2;
  border-radius: 100px;
  display: inline-block;
  height: 24px;
  position: relative;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  vertical-align: middle;
  width: 24px;
}

.emotion-1:after {
  background-color: #FFFFFF;
  border: 5px solid #FFFFFF;
  border-radius: 100px;
  content: "";
  display: block;
  height: 2px;
  margin-left: 6px;
  margin-top: 6px;
  opacity: 0;
  text-align: center;
  -webkit-transition: all .2s ease;
  transition: all .2s ease;
  width: 2px;
}

.emotion-2 {
  margin-left: 0.5rem;
}

.emotion-0 {
  opacity: 0;
  margin-right: -13px;
}

.emotion-0:hover + .emotion-1 {
  border-color: #7A7B8E;
}

.emotion-0:checked + .emotion-1 {
  background-color: #0069D0;
  border-color: #0069D0;
}

.emotion-0:checked + .emotion-1:after {
  opacity: 1;
}

.emotion-0:focus + .emotion-1 {
  box-shadow: 0 0 0 2px #F0F0F5,0 0 0 4px #0069D0;
}

.emotion-0:disabled + .emotion-1 {
  opacity: 0.4;
  cursor: not-allowed;
}

.emotion-0:disabled + .emotion-1 + .emotion-2 {
  color: #AAAAB9;
  cursor: not-allowed;
}

.emotion-25 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.emotion-4 {
  padding-top: 1rem;
  padding-left: 1rem;
  margin-bottom: 0.5rem;
  margin-right: 1rem;
}

<div>
  <div
    class="emotion-27"
  >
    <div
      class="emotion-26"
    >
      <div
        class="emotion-25"
      >
        <div
          class="emotion-4"
        >
          <label
            class="emotion-3"
          >
            <input
              checked=""
              class="emotion-0"
              name="aggregator"
              type="radio"
              value="0"
            />
            <span
              class="emotion-1"
            />
            <span
              class="emotion-2"
            >
              Count
            </span>
          </label>
        </div>
        <div
          class="emotion-4"
        >
          <label
            class="emotion-3"
          >
            <input
              class="emotion-0"
              name="aggregator"
              type="radio"
              value="1"
            />
            <span
              class="emotion-1"
            />
            <span
              class="emotion-2"
            >
              Percentage
            </span>
          </label>
        </div>
        <div
          class="emotion-4"
        >
          <label
            class="emotion-3"
          >
            <input
              class="emotion-0"
              name="aggregator"
              type="radio"
              value="2"
            />
            <span
              class="emotion-1"
            />
            <span
              class="emotion-2"
            >
              Average
            </span>
          </label>
        </div>
        <div
          class="emotion-4"
        >
          <label
            class="emotion-3"
          >
            <input
              class="emotion-0"
              name="aggregator"
              type="radio"
              value="3"
            />
            <span
              class="emotion-1"
            />
            <span
              class="emotion-2"
            >
              Maximum
            </span>
          </label>
        </div>
        <div
          class="emotion-4"
        >
          <label
            class="emotion-3"
          >
            <input
              class="emotion-0"
              name="aggregator"
              type="radio"
              value="4"
            />
            <span
              class="emotion-1"
            />
            <span
              class="emotion-2"
            >
              Minimum
            </span>
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
